### PR TITLE
logs: return application logs for dcs

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/origin/issues/4201

cc: @smarterclayton @deads2k @ironcladlou

### CODE REVIEW

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/registry/deploylog/rest.go [hunk 1]
```diff
@@ -173,3 +178,26 @@ func (g *podGetter) Get(ctx kapi.Context, name string) (runtime.Object, error) {
 	}
 	return g.podsNamespacer.Pods(namespace).Get(name)
 }
+
+// returnApplicationPodName tries to resolve the name for the oldest pod for the target deployment.
+func (r *REST) returnApplicationPodName(target *kapi.ReplicationController) (string, error) {
```
Comment (smarterclayton) [inline pkg/deploy/registry/deploylog/rest.go]: When we add the support proposed by Eric upstream, would the client side be responsible for doing the fan out?  I.e. would we have a Reaper equivalent (LogMulitplexer) for a DC and Deployment?  I assume it's not right to do it on the server, but makes it harder.

Comment (kargakis) [inline pkg/deploy/registry/deploylog/rest.go]: Not sure how this will be done

**Comment by smarterclayton** [comment]
Can you add a test here to end-to-end/core.sh so we can prove it works once something is deployed?  An extended test as well is not required unless you want to (can defer it to a future PR when we pull in the multi-item stuff)

**Comment by kargakis** [comment]
> Can you add a test here to end-to-end/core.sh so we can prove it works once something is deployed? An extended test as well is not required unless you want to (can defer it to a future PR when we pull in the multi-item stuff)

Added a test for application logs. I prefer to defer more testing for when we pull multiple logging

**Comment by kargakis** [comment]
new-app flakes
```
FAILURE after 30.922s: hack/../test/cmd/newapp.sh:14: executing 'oc new-app library/php mysql -o yaml' expecting success and text '3306': the command returned the wrong error code; the output content test failed
There was no output from the command.
Standard error from the command:
error: no match for "library/php", specify --allow-missing-images to use this image name.

The 'oc new-app' command will match arguments to the following types:

  1. Images tagged into image streams in the current project or the 'openshift' project
     - if you don't specify a tag, we'll add ':latest'
  2. Images in the Docker Hub, on remote registries, or on the local Docker engine
  3. Templates in the current project or the 'openshift' project
  4. Git repository URLs or local paths that point to Git repositories

--allow-missing-images can be used to point to an image that does not exist yet.

See 'oc new-app -h' for examples.
!!! Error in hack/../test/cmd/../../hack/cmd_util.sh:195
	'return 1' exited with status 1
Call stack:
	1: hack/../test/cmd/../../hack/cmd_util.sh:195 os::cmd::expect_success_and_text(...)
	2: hack/../test/cmd/newapp.sh:14 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:289
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:289 main(...)
Exiting with status 1
```
```
FAILURE after 31.370s: hack/../test/cmd/newapp.sh:15: executing 'oc new-app library/php mysql --dry-run' expecting success and text 'Image "mysql" runs as the 'root' user which may not be permitted by your cluster administrator': the command returned the wrong error code; the output content test failed
There was no output from the command.
Standard error from the command:
error: no match for "library/php", specify --allow-missing-images to use this image name.

The 'oc new-app' command will match arguments to the following types:

  1. Images tagged into image streams in the current project or the 'openshift' project
     - if you don't specify a tag, we'll add ':latest'
  2. Images in the Docker Hub, on remote registries, or on the local Docker engine
  3. Templates in the current project or the 'openshift' project
  4. Git repository URLs or local paths that point to Git repositories

--allow-missing-images can be used to point to an image that does not exist yet.

See 'oc new-app -h' for examples.
!!! Error in hack/../test/cmd/../../hack/cmd_util.sh:195
	'return 1' exited with status 1
Call stack:
	1: hack/../test/cmd/../../hack/cmd_util.sh:195 os::cmd::expect_success_and_text(...)
	2: hack/../test/cmd/newapp.sh:15 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:289
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:289 main(...)
Exiting with status 1
```
[test]

**Comment by kargakis** [comment]
ec2 failures

[test]

**Comment by kargakis** [comment]
[test]

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[merge]

**Comment by kargakis** [comment]
```
ERROR: gave up waiting for http://127.0.0.1:44001/version

!!! Error in hack/test-integration.sh:370
	'return 1' exited with status 1
Call stack:
	1: hack/test-integration.sh:370 main(...)
Exiting with status 1
```
[merge]

**Comment by kargakis** [comment]
https://github.com/openshift/origin/issues/6995 flake; test still hangs at the time of this post

**Comment by kargakis** [comment]
Removed following dc logs in e2e from deployments we know will be complete because following application logs isn't what we want to test (https://github.com/openshift/origin/issues/6995). cc: @mnagy @liggitt

**Comment by liggitt** [comment]
So this is an intentional behavior change that means that adding `--follow` to getting the logs of a deployment can become non-terminating if the deployment was already complete and you actually follow the application logs?

**Comment by kargakis** [comment]
> So this is an intentional behavior change that means that adding --follow to getting the logs of a deployment can become non-terminating if the deployment was already complete and you actually follow the application logs?

Exactly

**Comment by liggitt** [comment]
Seems like we'd hesitate to change behavior of existing commands to become arbitrarily non-terminating

**Comment by kargakis** [comment]
> Seems like we'd hesitate to change behavior of existing commands to become arbitrarily non-terminating

oc logs output is already non-terminating if you follow an application pod and people are still learning that we support dcs. Are you worried that somebody may flush their deployer logs in their scripts?

**Comment by liggitt** [comment]
it was a handy way to both capture the logs of a deployment and wait for it to finish. now it would behave differently if the deployment was already complete

**Comment by kargakis** [comment]
> **it was a handy way to both capture the logs of a deployment and wait for it to finish**. now it would behave differently if the deployment was already complete

I can't argue with that since I added those tests for this exact reason. @smarterclayton how do you feel about disabling -f for complete deployments? We will have it disabled either way for multiple logs.

**Comment by kargakis** [comment]
@liggitt what should be the behavior when trying to stream application logs? Error out or implicitly use --follow=false? Seems that we still break people in the first case and the second isn't desirable either, right?

**Comment by liggitt** [comment]
I don't know :-/

**Comment by smarterclayton** [comment]
-f for complete deployments displaying the log and then exiting I would
expect

**Comment by smarterclayton** [comment]
I would expect oc logs -f dc/foo to wait for the logs to finish on the
value of the current latest status version.

**Comment by kargakis** [comment]
> I would expect oc logs -f dc/foo to wait for the logs to finish on the value of the current latest status version.

IIUC, you mean that -f should stream like I already have it in this PR?

**Comment by smarterclayton** [comment]
I think if a deployment is running, oc logs -f is waiting for the
deployment to end.  If it's failed, you're showing the latest failed logs
(if they are found).  If its complete, you're showing the logs of the first
pod.

**Comment by kargakis** [comment]
The application logs are non-terminating (our app is supposedly running indefinitely). Someone who was previously running `oc logs -f dc/foo` might use the command not only to get the deployer logs but also as a mean to wait for the deployment process to finish.
```sh
# start a deployment
$ oc deploy foo --latest
# maybe do something intermediate
# wait for the process to finish while inspecting the logs at the same time
$ oc logs -f dc/foo
# do something dependent on a finished deployment; we know it's finished because oc logs -f terminated
```
 Now, if the process finishes earlier than `oc logs -f dc/foo` runs (can happen: https://github.com/openshift/origin/issues/6995) then oc logs will start streaming application (non-terminating) logs. So the flow I had before breaks. That's a concern to think about, though to be honest I prefer this than the alternatives (don't allow -f to work with complete deployments or implicitly default to --follow=false if we stream a complete deployment).

**Comment by smarterclayton** [comment]
I would expect oc deploy --follow to handle that behavior.

**Comment by kargakis** [comment]
@jwforres @spadgett fyi logs for a complete deployment return logs from the application pod now instead of just exiting.

**Pull request merged**